Add support for I2C EEPROMs with page sizes other than 128 in eeprom_…

adeuring opened the pull request:
Tinkering with AT24C64 EEPROMs and the micropython_eeprom library, I noticed errors, when I ran `full_test()` in `eeprom/i2c/eep_i2c.py`.

The reason: `EEPROM._getaddr()` used a fixed page size of 128 bytes to calculate the address of the next page and the length of the data than can be written in the next write transaction.

As documented for example in Atmel/Microchip datasheets, EEPROMs with smaller capacities use smaller page sizes, down to 32 bytes (AT24C32 and AT24C64):

https://ww1.microchip.com/downloads/en/devicedoc/doc0336.pdf
https://ww1.microchip.com/downloads/en/devicedoc/doc0670.pdf

This PR fixes the problem.

Two caveats:

1. While I added page sizes for the capacities 32kBit, 64kBit, 128kBit, 256kBit, 512kBit in the dictionary _PAGE_SIZES, I can currently test the changes with AT24C64 chips. (I have though ordered a small collection of breakout boards with EEPROMs of different capacities via Aliexpress. They should be delivered in two weeks or so.)

2. I am not very familiar with the market of I2C EEPROMs, hence I am not sure if the values of default page sizes in _PAGE_SIZES really make sense. In other words: Exist for example I2C EEPROMs with 32kBit capacity that have another pages size than 32?

Reason for the change in the test module `eep_i2c.py`: I tend to upload precompiled binaries to the Micropython device, and cptest() obviously fails in its current form when the device stores *.mpy files instead of the usual *.py ones.

adeuring commented:
Oh, the patch should fix these issues: #11, #20, #21

I had a look into the datasheets of the chips mentioned there, and they match the page sizes I assumed in my changes.

peterhinch reviewed (commented):
I get what you are aiming to do, but your `_PAGE_SIZES` table is changing the page size of chips that are already supported and have been tested. Please explain.

A better approach might be to have the constructor arg being `page_size=128` and documenting this, leaving it up to the user to determine the correct page size for unsupported chips.

adeuring commented:
I wrote the patch without looking into many datasheets of I2C EEPROMs and searching there for different pages sizes – I just assumed that EEPROMs with a given capacity from different manufacturers will most likely have the same page size, so I simply relied on the Microchup  datasheets linked above.

Out of curiosity, I searched Digikey for 64kBit I2C EEPROMs before writing this response. Digikey listed 395 different EEPROMs, with about 100 different links to datasheets.  (See the attached CSV file [digikey-i2c-eeproms-64k.csv](https://github.com/peterhinch/micropython_eeprom/files/13830598/digikey-i2c-eeproms-64k.csv) with the search result. Some links start with "//mm.digikey.com" – prefix these URLs with "https:".) 

Most datasheets mention a page size of 32 bytes.  Exceptions:
(1) https://ww1.microchip.com/downloads/en/DeviceDoc/21073K.pdf
and
https://rocelec.widen.net/view/pdf/asfxc3buan/CATSS01066-1.pdf
where the page size is 64.
(2) a few datasheets where I could not find anything about the page size: 
https://mm.digikey.com/Volume0/opasdata/d220001/medias/docus/1005/br24cxx-xx6tp-e.pdf
https://mm.digikey.com/Volume0/opasdata/d220001/medias/docus/9/br24cxx-wxx6tp-e.pdf
https://www.rohm.de/datasheet?p=BR24T64-WZ&dist=Digi-key&media=referral&source=digi-key.com&campaign=Digi-key
(It is of course possible that I simply was too dumb to find the place in  these PDFs where the page size is mentioned....)

So I think that at least the entry for 64kBit EEPROMs in `_PAGE_SIZES` makes sense.

That said, I do not insist to keep the patch as it is. I can of course change the call parameter `page_size=None` to `page_size=128` and drop `_PAGE_SIZES`. On the other hand: When `class EEPROM` operates with a too large page size, corrupted data will be written, as can be seen in at least three bug reports. But when it operates with a smaller page size than possible, write operations simply become slower than necessary.

peterhinch commented:
> when it operates with a smaller page size than possible, write operations simply become slower than necessary.

Fair point, but I'm reluctant to change working code. The supported device types were tested.

It's also worth bearing in mind that users of this code are working at chip level and can be assumed to be fairly experienced. I think we should proceed as per my suggestion, but document `page_size` with your above observation. If corruption occurs, or the user is cautious by nature, a lower value can be set.

I also think that trying to second-guess the vast range of devices out there is rather ambitious: let the user do the research.

Should there be a check that `page_size` is an integer power of 2?

adeuring commented:
OK, I've changed the "meaning" of constructor parameter `page_size`  from the page size itself to the binary log of the page size. Makes the parameter indeed "more similar" to `block_size` and "saves" a consistency check that the value is a power of 2.

I also changed the default value of `page_size` to 7, so the behaviour of the constructor is the same as before when no value is specified. But I am still not 100% convinced: I'd really like to see the datasheet of an I2C EEPROM with capacity of at most 256kBit that has a page size of 128.

And I updated the doc. May very well contain typos and I'd welcome suggestions for improvements.

peterhinch reviewed (approved):
That looks good!

peterhinch commented:
Your changes to the driver test fine with `test()` and `full_test()` passing on my test rig. But the changes to `cptest()` aren't working here.
```python
>>> eep_i2c.cptest()
2 chips detected. Total EEPROM size 131072bytes.
Instantiated EEPROM
Device already mounted.
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/sd/eep_i2c.py", line 138, in cptest
  File "/sd/eep_i2c.py", line 26, in cp
OSError: [Errno 2] ENOENT
```
I then did this:
```python
>>> import os
>>> os.listdir('/')
['flash', 'sd', 'eeprom']
>>> os.listdir('/eeprom/')
['eep_i2c.py']
>>> 
```

adeuring commented:
> [EDIT] This may have been a function of my test setup. I have pushed a patch to cptest: rather than finding the suffix, it does a `replace` on the fiename. This preserves the path to the file.

Right, that maked sense. I tried to reproduce the error with my setup and got stuck with another error: Copying the *py files failed with `OSError: 28` in the write() call. That error number is unfortunately not part ot the rrno module; the constant is defined in `lib/littlefs/lfs2.h`:

```
    LFS2_ERR_NOSPC       = -28,  // No space left on device
```

Makes sense when you try to copy two files with a length of about 4000 bytes each to an 8kByte device.

> Regarding page sizes, I do take your point. It does seem there has been a bug lurking for several years and it puzzles me that the smaller devices worked. As you point out, the only drawback of setting a smaller page size is speed so maybe the default should be set to 5 (32 bytes)?

Looks reasonable to me, from digging around in the datasheets of 64kBit EEPROMs. Several of these datasheets document also 32kBit EEPROMs with the same page size (32). Smaller EEPROMs have smaller page sizes, but they are not easily supported anyway by `class EEPROM`, since they use only one "real" address byte. For EEPROMs with a capacity of more that 256 bytes but at most 2048 bytes, the higher address bits - up to three - are stuffed into the lower bits of the I2C address... So the second address byte, as written by `class EEPROM`, would be treated as the first data byte to be written by these smaller EEPROMs. (Oh, BTW, EEPROMs with a capacity of more than 64kByte use the same trick, see for example https://ww1.microchip.com/downloads/en/devicedoc/doc1471.pdf). The nice thing about this abuse of I2C addresses: These chips should work fine with `class EEPROM` in its current form, when `max_chips_count` is 
 2 or larger.

> 
> It might be interesting to write a test (in eep_i2c.py) which measured the page size. Would the following work?
> 
> ```python
> eep = EEPROM(<args>, page_size=7)
> arr= bytearray(256)
> for x in range(256):
>     eep[x] = 255  # Write single bytes, don't invoke page write
> eep[0:256] = arr  # Zero 256 bytes attempting to use 128 byte pages
> for x in range(256):
>     if eep[x]:
>         print(f"Page size = {x}")
>         break
> ```
> 
> I can't try this unless I can find a chip with page size < 128 - I'll have a dig around tomorrow.

Looks good; I tried it with an AT24C64; the result (32) is what the datasheet says. Just one suggestion: Use an larger address range. The AT24C1024 has a page size of 256. Perhaps it makes sense to derive the address range for the test from the capacity of the chip, as declared by the parameter `chip_size`.

peterhinch commented:
Thanks for testing the above - point taken re 256 byte page size. I plan to give this my full attention over the next few days. Tasks:
* Investigate SPI devices too.
* Quantify the speed loss of specifying a page size of 32 on devices with larger physical page sizes (especially on SPI).
* Consider measuring the page size in the constructor. This would avoid the need for the `page_size` arg.

adeuring commented:
> * Consider measuring the page size in the constructor. This would avoid the need for the `page_size` arg.

I think that it is risky to do this in the constructor of class EEPROM. At least, if this happens in each __init__() call in the "normal operation mode" of an application. Determining the page size requires to (over)write a few bytes in the EEPROM, so there is the risk of data corruption. Sure, you can keep a copy of the data block that will be used for testing in RAM and restore it at the end of the test. But what if power is lost during before the data is restored?

I think a dedicated function to determine the page size, as you you sketched above, is a much better approach, combined with a warning in the documentation that calling the function will overwrite some part of the EEPROM.

peterhinch commented:
The solution I have come up with is as follows. There is a new base class in `bdevice.py`:
```python
class EepromDevice(BlockDevice):
    def __init__(self, nbits, nchips, chip_size, page_size, verbose):
```
This is shared between I2C and SPI classes and handles page size.

The `EEPROM` constructor still has the `page_size` arg which now defaults to `None`. In this case the page size is determined by the new base class. The relevant device contents are saved to RAM and restored. I have improved the speed of this: on my hardware it takes ~50ms. However if `page_size` is set to an integer, it is taken as correct with no hardware check (although there is validation on passed values). This option saves time, RAM allocations and write cycles and enables users to avoid power loss risks.

The test script `eep_i2c.test()` now explicitly checks for a `page_size` value that exceeds the physical page size.

I now need to sort out the SPI case, although with a shared base class this should be straightforward.

adeuring commented:
> The `EEPROM` constructor still has the `page_size` arg which now defaults to `None`. In this case the page size is determined by the new base class. The relevant device contents are saved to RAM and restored.  I have improved the speed of this: on my hardware it takes ~50ms. However if `page_size` is set to an integer, it is taken as correct with no hardware check (although there is validation on passed values). This option saves time, RAM allocations and write cycles and enables users to avoid power loss risks.

Sounds good, But I think there should be an explicit warning about the risks of the page size auto detection in the docs. (I am somewhat paranoid ;))

peterhinch commented:
My plan to fix SPI has been delayed because of hardware problems with my test rig. I'll have to roll my sleeves up and fix it.

Meanwhile I've pushed an update. This includes all the I2C work. I have amended the docs, including a warning about auto detection.

Comments welcome.

adeuring commented:
> Meanwhile I've pushed an update. This includes all the I2C work. I have amended the docs, including a warning about auto detection.
> 
> Comments welcome.

Purely from reading the diff – I haven't tried it yet: Looks really good, just one remark:

I would put the lines to autodetect the page size, i.e., 103 to 113 in bdevice.py, into a try/except block, like so:
```
try:
    self._psize(256)
    ...
    self[:129] = old
except:
    for i, v in enumerate(old):
       self[i] = v
```
As already mentioned elsewhere, I'm a bit paranoid ;) If anything except hardware failures goes wrong during the page size auto detection, this will still restore the EEPROM contents.

peterhinch commented:
Now done.

peterhinch commented:
I'm having second thoughts about this. Determining page size is now easy: run `test()` which uses auto detect and verbose. Page size is reported. In production code, set `page_size=<something>` and the hazard is avoided completely.

peterhinch commented:
Re power outages.

The SPI driver was written to support two specific chips and used proprietary means of detecting chip presence - these differed between the two chips. These had the benefit of leaving the data intact and therefore being robust in the event of an outage.

It has become clear from you and others that there is a wish to support a wider range of chips. I have therefore adapted the SPI driver to eliminate proprietary commands. Presence detection is done by performing a readback check on an array location. This carries a risk if an outage occurs. As discussed, this also applies to page size detection. I intend to handle this as follows.

The docs will now make this risk clear, and advise that production code specifies the page size. On SPI arrays presence detection is only performed if `verbose=True`. On production code the docs will now recommend setting this `False`.

It is now easy to determine page size empirically and to explicitly specify it. There is also no reason to perform presence detection in a production system. Its role is in test setups. In a production system it could only ever have detected a catastrophic chip failure.

I think this addresses the issue, but comments are welcome.
